### DISCUSSION TITLE
Support for Kubernetes clusters

### DISCUSSION THREAD

**Discussion by digiserg** [axonops/axonops-workbench#319] [Ideas]
We could add support for connecting to Cassandra clusters running in Kubernetes either via Ingress, HostPort of Kube-Proxy.

**Comment by millerjp** [comment]
@digiserg could you elaborate a bit more on how this could possibly work natively with in the application

**Reply by digiserg** [reply]
Sure! I personally think that support for [port-forwarding](https://kubernetes.io/docs/tasks/access-application-cluster/port-forward-access-application-cluster/) would be the nicest one to have. 

The configuration window for the Workbench would be similar to using SSH forwarding but the user would need to provide instead:

- Path to KUBECONFIG file
- Path to the `kubectl` cli if it cannot be detected
- Namespace where the Cassandra cluster is running
- Service name (this could be drop down list that can be discovered)

Workbench would then start up a port-forward like

```sh
kubectl --kubeconfig PATH_TO port-forward -n NAMESPACE svc/CASSANDRA_SERVICE_NAME RANDOM_LOCAL_PORT:9042
```

This will provide a local port `RANDOM_LOCAL_PORT` for the `cqlsh` command to connect to.

There may be some nodejs lib to this natively instead of using the `kubectl` cli but I'm not aware of any.

**Reply by millerjp** [reply]
Would https://github.com/mittwald/node-kubernetes provide the necessary support? I am not sure they even need to have kubectl locally to use this, it just interacts with the rest APIs?

See: https://www.perplexity.ai/search/if-i-wanted-to-setup-a-portfor-DPl1yEGDTgWamtqIqhVZPw

**Reply by digiserg** [reply]
The example you sent is actually using `kubectl`, the library does not support port-forward. It can be used for the discovery phase though. For example, to list all the services in the namespace and present a drop down to the user.

**Comment by millerjp** [comment]
@digiserg would this need to run over an ssh tunnel also i.e would a desktop app have direct access to the k8 rest APIs?

**Reply by digiserg** [reply]
Both probably, yes. As a user, I'm likely to have direct access to some k8s clusters but require a bastion for others.
